overlord/hookstate/ctlcmd/get.go

```diff
@@ -120,3 +135,126 @@ func (c *getCommand) Execute(args []string) error {
+	}
+
+	if c.Typed && c.Document {
+		return fmt.Errorf("cannot use -d and -t together")
+	}
+
+	var snapAndPlugOrSlot snap.SnapAndName
+	// treat PlugOrSlotSpec argument as config key if it doesn't contain ':'
+	if !strings.Contains(c.Positional.PlugOrSlotSpec, ":") {
+		c.Positional.Keys = append([]string{c.Positional.PlugOrSlotSpec}, c.Positional.Keys[0:]...)
+		c.Positional.PlugOrSlotSpec = ""
+	} else {
+		snapAndPlugOrSlot.UnmarshalFlag(c.Positional.PlugOrSlotSpec)
+	}
+
+	if snapAndPlugOrSlot.Name != "" {
+		return c.handleGetInterfaceAttributes(context, snapAndPlugOrSlot.Snap, snapAndPlugOrSlot.Name)
+	}
+
+	if c.ForcePlugSide || c.ForceSlotSide {
+		return fmt.Errorf("cannot use --plug or --slot without <snap>:<plug|slot> argument")
+	}
+
+	context.Lock()
+	transaction := configstate.ContextTransaction(context)
+	context.Unlock()
+
+	return c.printValues(func(key string) (interface{}, bool, error) {
+		var value interface{}
+		err := transaction.Get(c.context().SnapName(), key, &value)
+		if err == nil {
+			return value, true, nil
+		}
+		if configstate.IsNoOption(err) {
+			if !c.Typed {
+				value = ""
+			}
+			return value, false, nil
+		}
+		return value, false, err
+	})
+}
+
+func (c *getCommand) handleGetInterfaceAttributes(context *hookstate.Context, snapName string, plugOrSlot string) error {
+	// Make sure get :<plug|slot> is only supported during the execution of prepare-[plug|slot] hooks
+	var isPreparePlugHook, isPrepareSlotHook, isConnectPlugHook, isConnectSlotHook bool
+
+	if strings.HasPrefix(context.HookName(), "prepare-plug-") {
+		isPreparePlugHook = true
+	} else if strings.HasPrefix(context.HookName(), "connect-plug-") {
+		isConnectPlugHook = true
+	} else if strings.HasPrefix(context.HookName(), "prepare-slot-") {
+		isPrepareSlotHook = true
+	} else if strings.HasPrefix(context.HookName(), "connect-slot-") {
+		isConnectSlotHook = true
+	}
+	if !(isPreparePlugHook || isPrepareSlotHook || isConnectPlugHook || isConnectSlotHook) {
+		return fmt.Errorf(i18n.G("interface attributes can only be read during the execution of interface hooks"))
+	}
+
+	var err error
+	var attributes map[string]map[string]interface{}
+
+	context.Lock()
+	defer context.Unlock()
+	err = context.Get("attributes", &attributes)
+
+	if err == state.ErrNoState {
+		return fmt.Errorf(i18n.G("attributes not found"))
+	}
+	if err != nil {
+		return err
+	}
+
+	if c.ForcePlugSide && c.ForceSlotSide {
+		return fmt.Errorf("cannot use --plug and --slot together")
+	}
+
+	// the typical case, we don't expect snap name to be provided via snapctl get :<plug|slot> ...
+	// if it's provided it should be the current snap, otherwise it's an error.
+	if snapName == "" {
+		isPlugSide := (isPreparePlugHook || isConnectPlugHook)
+		isSlotSide := (isPrepareSlotHook || isConnectSlotHook)
+		if (isSlotSide && c.ForcePlugSide) || (isPlugSide && c.ForceSlotSide) {
+			// get attributes of the remote end
+			err = context.Get("other-snap", &snapName)
```

Comment:
Let's please talk about the storage in context. Would just like a guided tour through the current thinking around how the data is stored (attributes map, other-snap, etc), and potential alternatives considered.